[Refactor] refactor some code for Cluster Snapshot

srlch opened the pull request:
## What I'm doing:
refactor some code for Cluster Snapshot:
1. make some function interfaces more flexible
2. remove write/read override for ClusterSnapshotJob, because it is useless.
3. Explictly wait for the loop for ClusterSnapshotCheckpointScheduler 

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshot.java:
```diff
@@ -45,23 +45,33 @@ public enum ClusterSnapshotType {
         this.starMgrJournalId = starMgrJournalId;
     }
 
+    public ClusterSnapshot(long id, String snapshotName, String storageVolumeName, long createdTimeMs,
+                           long finishedTimeMs, long feJournalId, long starMgrJournalId) {
+        this(id, snapshotName, ClusterSnapshotType.AUTOMATED, storageVolumeName, createdTimeMs, finishedTimeMs,
+                feJournalId, starMgrJournalId);
+    }
```
only omitted snapshot type, this construct function could be removed ?

srlch replied:
fixed, remove this constructor

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotJob.java:
```diff
@@ -56,7 +50,9 @@ public enum ClusterSnapshotJobState {
+        if (snapshotName.startsWith(ClusterSnapshotMgr.AUTOMATED_NAME_PREFIX)) {
+            this.snapshot = new ClusterSnapshot(id, snapshotName, storageVolumeName, createdTimeMs, -1, 0, 0);
+        }
         this.state = ClusterSnapshotJobState.INITIALIZING;
```
do not use snapshot name to tell whther it a automated snapshot, add a snapshot type arg

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotJob.java:
```diff
@@ -56,12 +50,18 @@ public enum ClusterSnapshotJobState {
         this.detailInfo = "";
     }
 
+    public ClusterSnapshot newClusterSnapshot(long id, String snapshotName, String storageVolumeName, long createdTimeMs) {
```
```suggestion
    protected ClusterSnapshot newClusterSnapshot(long id, String snapshotName, String storageVolumeName, long createdTimeMs) {
```

srlch replied:
updated

alvin-celerdata commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -127,6 +127,19 @@ protected void clearFinishedAutomatedClusterSnapshot(String keepSnapshotName) {
+        }
+
+        try {
+            Thread.sleep(Config.automated_cluster_snapshot_interval_seconds * 1000L);
```
The automated_cluster_snapshot_interval_seconds default value is 600. It isn't a good design to let a function caller wait for 600 seconds. Many things can happen during that time, but the callers can do nothing except hang up and wait for the time to pass.

srlch replied:
refactored the code, control interval in `ClusterSnapshotCheckpointScheduler.runAfterCatalogReady` by timestamp instead of Daemon framework or function call in `ClusterSnapshotMgr`.

xiangguangyxg replied:
Here I think this function could add an argument `lastJobStartTimeMs` to make it return a new automated job in a certain time.

srlch committed:
[Refactor] refactor some code for Cluster Snapshot

## What I'm doing:
refactor some code for Cluster Snapshot:
1. make some function interfaces more flexible
2. remove write/read override for ClusterSnapshotJob, because it is useless.
3. Explictly wait for the loop for ClusterSnapshotCheckpointScheduler

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -127,6 +127,19 @@ protected void clearFinishedAutomatedClusterSnapshot(String keepSnapshotName) {
+        try {
+            Thread.sleep(Config.automated_cluster_snapshot_interval_seconds * 1000L);
+        } catch (InterruptedException e) {
+            LOG.error("InterruptedException: ", e);
```
refactored the code

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -65,6 +60,22 @@ protected void runAfterCatalogReady() {
+            return;
+        }
+
+        lastStartTimeMs = System.currentTimeMillis();
```
Fixed

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -65,6 +60,22 @@ protected void runAfterCatalogReady() {
+
+        lastStartTimeMs = System.currentTimeMillis();
+        runningJob = GlobalStateMgr.getCurrentState().getClusterSnapshotMgr().createAutomatedSnapshotJob();
+        if (runningJob == null) {
```
Fixed, move it into try block instead.

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotJob.java:
```diff
@@ -79,6 +78,10 @@ public void setErrMsg(String errMsg) {
 
+    public void setCreatedTimeMs(long createdTimeMs) {
+        snapshot.setCreatedTimeMs(createdTimeMs);
+    }
```
Removed this function, and keep created timestamp set only in constructor function.

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -30,53 +29,59 @@ public class ClusterSnapshotCheckpointScheduler extends FrontendDaemon {
+    protected final RestoredSnapshotInfo restoredSnapshotInfo;
 
-    private boolean firstRun;
+    protected boolean firstRun;
```
done

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotCheckpointScheduler.java:
```diff
@@ -30,53 +29,59 @@ public class ClusterSnapshotCheckpointScheduler extends FrontendDaemon {
+            runningJob = GlobalStateMgr.getCurrentState().getClusterSnapshotMgr().getNextCluterSnapshotJob();
+
+            // set last start time when job has been created and begin to submit
+            lastAutomatedJobStartTimeMs = System.currentTimeMillis();
```
done

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:
```diff
@@ -127,6 +127,15 @@ protected void clearFinishedAutomatedClusterSnapshot(String keepSnapshotName) {
+                                           Config.automated_cluster_snapshot_interval_seconds * 1000L);
+    }
+
+    public ClusterSnapshotJob getNextCluterSnapshotJob() {
+        return createAutomatedSnapshotJob();
+    }
+
     public ClusterSnapshotJob createAutomatedSnapshotJob() {
```
Because we want to override `getNextCluterSnapshotJob` in the future pr.

alvin-celerdata replied:
Please add an override annotation

srlch replied:
This is a base class, it seems that i can not add an override annotation?
